### PR TITLE
[Loader] Serializer: Add GC proposal types

### PR DESCRIPTION
**Pull request by elhewaty**
Signed-off-by: Mohamed Atef <mohamedatef1698@gmail.com>

### CODE REVIEW

**Comment by elhewaty** [comment]
@q82419 @hydai I will add the tests but PTAL.
another thing is: did you handle [Defined Types](https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#defined-types), as I don't find handling for them in the loader.
and I didn't find support for the `rec` type in the loader. I think I am missing something.

**Comment by q82419** [comment]
> @q82419 @hydai I will add the tests but PTAL. another thing is: did you handle [Defined Types](https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#defined-types), as I don't find handling for them in the loader. and I didn't find support for the `rec` type in the loader. I think I am missing something.

I'll take a look in these days. You can keep adding the tests.

The defined type and recursive type are very complicated.
Defined types are the types defined in the module, that is, the types in `TypeSection`, which is `vec(rectype)`.
You can find the binary format of recursive type [here](https://webassembly.github.io/gc/core/binary/types.html#recursive-types).
Recursive type is a vector of subtype or a single subtype.
In our implementation, we expand the recursive types and store the subtype list into the type section, with the recursive type group information.
You can find that we handle the expanding of recursive type group in `lib/loader/ast/section.cpp`.

**Comment by elhewaty** [comment]
@ q82419, here is a snippet:

```
TEST(serializeTypeTest, SerializeGCType) {
  std::vector<uint8_t> Output;
  std::vector<uint8_t> Expected;

  // 6. Test serialize GC heap types.
  //
  //   1. Test NullFuncRef type.
  //   ...

  WasmEdge::AST::GlobalType GlobalType;
  Conf.addProposal(WasmEdge::Proposal::GC);

  GlobalType.setValType(WasmEdge::TypeCode::NullFuncRef);
  GlobalType.setValMut(WasmEdge::ValMut::Const);
  Output = {};
  EXPECT_TRUE(Ser.serializeSection(createGlobalSec(GlobalType), Output));

  Expected = {
      0x06U, // Global section
      0x04U, // Content size = 4
      0x01U, // Vector length = 1
      0x73U, // NullFuncRef type, why is this wrong? 0xF3 0x00 is correct?
      0x00U, // Const mutation
      0x0BU  // Expression
  };
  EXPECT_EQ(Output, Expected);
...
}
```
in GC NullFuncRef appears as a heap type, ref type, and a valtype, with the same opcode. and I think they all handled the same with [here](https://github.com/WasmEdge/WasmEdge/blob/master/include/common/types.h#L120)  so one test is efficient -for every type- right? or there's a difference in handling these type in ref, heap or valtype types?

one more thing is:
if we replace `GlobalType.setValType(WasmEdge::TypeCode::NullFuncRef);` above with `GlobalType.setValType(WasmEdge::TypeCode::I8);` `Expected` becomes `{6, 1}`!!

I think I missed many things here! given this test case and the above code what did I do wrong ?? (a lot I know, but sorry:cry:)

Sorry for all these questions

**Comment by elhewaty** [comment]
@q82419 
[here](https://webassembly.github.io/gc/core/syntax/types.html#syntax-arraytype) `fieldtype :: mut storagetype` but in the github proposal `field type :: storagetype mut` https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#field-types

**Comment by q82419** [comment]
> @q82419 [here](https://webassembly.github.io/gc/core/syntax/types.html#syntax-arraytype) `fieldtype :: mut storagetype` but in the github proposal `field type :: storagetype mut` https://github.com/WebAssembly/gc/blob/main/proposals/gc/MVP.md#field-types

@elhewaty 
There may be unsync cases.
Please follow the modified spec.

**Comment by q82419** [comment]
> @q82419 Oh thanks, I follow it but I think I followed a lot of links so I ended in the wasm structure page. I think the PR is ready for review.

I'll review in these days. Thanks!

**Comment by elhewaty** [comment]
I will open another PR for instructions. Should I wate for this to land, or add to this incrementally?

**Comment by q82419** [comment]
> I will open another PR for instructions. Should I wate for this to land, or add to this incrementally?

You can add commits into this PR.

1. Please rebase to the master. We fixed some building issues.
2. Please fix the build error on Windows.

**Comment by elhewaty** [comment]
@q82419 I added more tests to improve the test coverage and fixed the build error.
once this is merged I will create a separate PR for instructions.

**Comment by elhewaty** [comment]
@q82419 ping

**Comment by q82419** [comment]
> @q82419 ping

Sorry that I'm busy in fixing other issues.
Please help to keep improve the coverage, and I'll review it in the following days. Thanks!

File: lib/loader/serialize/serial_type.cpp
Comment (q82419) [inline lib/loader/serialize/serial_type.cpp:120]: Not need the following cases.
Because in the `class ValType`, we migrated the ref-related cases into the `ref` or `ref.null`, and handled in the heap type.

File: lib/loader/serialize/serial_type.cpp
Comment (q82419) [inline lib/loader/serialize/serial_type.cpp:72]: Not need the following cases.
Because in the `class ValType`, we migrated the ref-related cases into the `ref` or `ref.null`, and handled in the heap type.

**Review by q82419 — CHANGES_REQUESTED** [review]
Please complete your implementation with following:

1. You only implemented the composite types, which have the same scope as functype originally. Please complete the [recursive types](https://webassembly.github.io/gc/core/binary/types.html#recursive-types).
2. It's OK for using the global type as the ValType testing. But it will be lack of field type tests.

**Comment by elhewaty** [comment]
@q82419  I am working on adding instructions, I am adding the tests. Regarding the recursive types, aren't they handlee [here](https://github.com/WasmEdge/WasmEdge/pull/3072)

**Comment by q82419** [comment]
> @q82419 I am working on adding instructions, I am adding the tests. Regarding the recursive types, aren't they handlee [here](https://github.com/WasmEdge/WasmEdge/pull/3072)

Looks like I implemented before.
Can you help to add the unit tests for the coverage?
You may need to hand-write some binaries.
For some tests, you can refer to our [spec test repo](https://github.com/second-state/WasmEdge-unittest/tree/wasm-dev-0.14.0/gc). This repo expand the `wast` files of spec test into the WASM binaries.
Thanks!

**Comment by elhewaty** [comment]
Of course I will add more tests. Can you please give an example of hand-written binaries? I can't open the files. May be give some more details.

Thanks

**Comment by q82419** [comment]
> Of course I will add more tests. Can you please give an example of hand-written binaries? I can't open the files. May be give some more details.
> 
> Thanks

They are the binary files, so you should use the hex editor to open them or read the binary from some codes.
Then you can create some binary examples as you did with the global section:

```
std::vector<uint8_t> WasmBinary = {
  0x00U, 0x61U, 0x73U, 0x6DU, // Magic
  0x01U, 0x00U, 0x00U, 0x00U, // Version
  0x01U,                                     // Type section
  0x04U,                                     // Content size = 4
  0x01U,                                     // Vector length = 1
  0x5EU,                                     // Array type
  0x78U,                                     // Storage type: i8
  0x01U,                                     // Mutability: mut
};
```

**Comment by elhewaty** [comment]
@q82419 BTW, how to be one of the wasmedgd maintainers?

**Comment by q82419** [comment]
> @q82419 BTW, how to be one of the wasmedgd maintainers?

I think @hydai is more clearly about this question.

**Comment by hydai** [comment]
> @q82419 BTW, how to be one of the wasmedgd maintainers? 

Please check the section, Adding maintainers in 
https://github.com/WasmEdge/WasmEdge/blob/master/docs/GOVERNANCE.md

Before a person is nominated as a maintainer, they will be a reviewer and committer first. Once they continue to contribute to significant features, they will be nominated as a reviewer or committer. And for a long-term contribution, they will be nominated as a new maintainer.

**Comment by elhewaty** [comment]
@q82419, can you take a look at this? The only thing remaining is the `rectype` test.

File: lib/loader/serialize/serial_instruction.cpp
Comment (elhewaty) [inline lib/loader/serialize/serial_instruction.cpp:196]: @q82419 Can you help me with this case? 
the implementation is wrong I think, I get segmentation fault while testing. any hint?

File: lib/loader/serialize/serial_instruction.cpp
Comment (q82419) [inline lib/loader/serialize/serial_instruction.cpp:152]: This is incorrect.
According to the [binrary format](https://webassembly.github.io/gc/core/binary/instructions.html#reference-instructions) of `ref.test` and `ref.cast` instructions, the immediates are the `heaptype`, not `reftype`.
You can refer to our [implementation here](https://github.com/WasmEdge/WasmEdge/blob/master/lib/loader/ast/instruction.cpp#L520-L538).
We loaded these instructions with their `heaptype` immediates, and set into the `valtype` class.
Therefore, for the serialization, you should output the [`heaptype`](https://webassembly.github.io/gc/core/binary/types.html#binary-heaptype) without the `ref` or `ref.null` decoration.

File: lib/loader/serialize/serial_instruction.cpp
Comment (q82419) [inline lib/loader/serialize/serial_instruction.cpp:202]: May be `RType2`?

File: lib/loader/serialize/serial_instruction.cpp
Comment (q82419) [inline lib/loader/serialize/serial_instruction.cpp:209]: As above, `heaptype` here.

File: lib/loader/serialize/serial_instruction.cpp
Comment (q82419) [inline lib/loader/serialize/serial_instruction.cpp:215]: As above, `heaptype` here.

File: lib/loader/serialize/serial_type.cpp
Comment (q82419) [reply lib/loader/serialize/serial_type.cpp]: Not fixed.

File: lib/loader/serialize/serial_type.cpp
Comment (q82419) [reply lib/loader/serialize/serial_type.cpp]: Not fixed.

File: lib/loader/serialize/serial_type.cpp
Comment (q82419) [inline lib/loader/serialize/serial_type.cpp:72]: As commented before, the following cases will always not be reached because the `getCode()` for `reftype` will always be `TypeCode::Ref` or `TypeCode::RefNull`.

File: lib/loader/serialize/serial_type.cpp
Comment (q82419) [inline lib/loader/serialize/serial_type.cpp:119]: As commented before, the following cases will always not be reached because the `getCode()` for `valtype` will always be the cases above.
REF: https://github.com/WasmEdge/WasmEdge/blob/master/include/common/types.h#L63-L85

**Review by q82419 — CHANGES_REQUESTED** [review]
You should have a more understanding about reference types in GC:

1. `heaptype` is the [binrary format defined here](https://webassembly.github.io/gc/core/binary/types.html#heap-types).
2. `reftype` is the `heaptype` with `ref` or `ref.null` decoration, or `heaptype` itself, [binary format defined here](https://webassembly.github.io/gc/core/binary/types.html#reference-types). In WasmEdge, we migrated the all `heaptype` cases into `ref.null heaptype`, that is, only `TypeCode::Ref` and `TypeCode::RefNull` cases in a `valtype` class. When serializing, you should focus on whether output the `ref` (0x64) or `ref.null` (0x63) decoration or not.
3. `valtype` is the largest super set. It contains the `vectype`, `numtype`, and `reftype`. As above mentioned in `reftype`, the `valtype` only contains `TypeCode::Ref` and `TypeCode::RefNull` cases when it's a `reftype`.

**Comment by elhewaty** [comment]
@q82419 ping

File: test/loader/serializeInstructionTest.cpp
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:1246]: These instructions are also reference instrucitons. Combine them into the test set.

File: test/loader/serializeInstructionTest.cpp
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:1289]: These instructions are also reference instrucitons. Combine them into the test set.

File: test/loader/serializeInstructionTest.cpp
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:1330]: These instructions are also reference instrucitons. Combine them into the test set.

File: test/loader/serializeInstructionTest.cpp
Comment (q82419) [inline test/loader/serializeInstructionTest.cpp:1425]: These instructions are also reference instrucitons. Combine them into the test set.

File: test/loader/serializeTypeTest.cpp
Comment (q82419) [inline test/loader/serializeTypeTest.cpp:365]: I think you can combine this with the `SerializeFuncRefType` test, and rename as `SerializeValType`.
After all they are all tests about serializing the value types.
Just take care about the proposal configuration.

File: test/loader/serializeTypeTest.cpp
Comment (q82419) [inline test/loader/serializeTypeTest.cpp:371]: You only added the abstract heap type with `ref.null` cases.
Add some tests about the following case for example:

```
GlobalType.setValType(WasmEdge::TypeCode::Ref, WasmEdge::TypeCode::StructRef);
// This will output `0x64U, 0x6BU` (ref struct)

GlobalType.setValType(WasmEdge::TypeCode::RefNull, WasmEdge::TypeCode::StructRef);
// This will output `0x6BU` (ref.null struct), 
// which is the same as `setValType(WasmEdge::TypeCode::StructRef)`
// to prove that your code outputed the shorten reftype.
```

**Review by q82419 — CHANGES_REQUESTED** [review]
Almost done about the tests of instructions and value types.
You can add tests about recursive types, sub types, and field types in this or another PR.

**Comment by elhewaty** [comment]
Thanks for reviewing, I will update the patch today.

**Comment by elhewaty** [comment]
@q82419  I added some RecType, subtype, Composite, and fieldtype tests should I update this PR?

**Comment by q82419** [comment]
> @q82419 I added some RecType, subtype, Composite, and fieldtype tests should I update this PR?

I think this PR is ready to merge.
You can create another PR, thanks!
Reminder that record these PRs into your issue.

**Review by q82419 — APPROVED** [review]

**Pull request merged**
